pkg/csi/service/common/authmanager.go

```diff
@@ -0,0 +1,179 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package common
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/vmware/govmomi/object"
+	"github.com/vmware/govmomi/vim25/types"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"sigs.k8s.io/vsphere-csi-driver/pkg/common/cns-lib/vsphere"
+	cnsvsphere "sigs.k8s.io/vsphere-csi-driver/pkg/common/cns-lib/vsphere"
+	"sigs.k8s.io/vsphere-csi-driver/pkg/csi/service/logger"
+)
+
+// AuthorizationService exposes interfaces to support authorization check on datastores and get datastores which need to be
+// ignored by create volume
+type AuthorizationService interface {
+
+	// GetDatastoreIgnoreMapForBlockVolumes returns a map which maps datastore url to datastore info which user does not have Datastore.FileManagement privilege
+	// returns an error if not able to retrieve the map
+	GetDatastoreIgnoreMapForBlockVolumes(ctx context.Context) (map[string]*cnsvsphere.DatastoreInfo, error)
```

Comment:
Maintaining an ignore map on AuthorizationService assumes that the caller of this method always to have list of datastores and compare it with ignore map to remove those from that full list. 

Why not go with a approach where we maintain datastores map only for those datastores which user has permissions on.


This way `GetDatastoreIgnoreMapForBlockVolumes` will be renamed to `GetDatastoreMapForBlockVolumes`

Reply:
I have discussed with @SandeepPissay offline on this and he seems to be fine with this approach. Since this is a simple change can you take care of it.

Reply:
For the current use case, create volume has a list of datastores to compare.  If you only maintains datastores which user has permission, you need to do the intesection of two set of datastores.

For example, for create volume, there are two shared datastores, datastore1 and datastore2. Only datastore1 matches the given storage policy.  sharedDatastore={datastore1}
Both datastore1 and datastore2 has the privilege for creating volume.  GetDatastoreMapForBlockVolume = {datastore1, datastore2}
You need do the intesection to find out the datastore which need to passed to CNS backend.

Reply:
I will not make this change.

Reply:
I think you got it wrong. In your example, 
GetDatastoreMapForBlockVolume = {datastore1, datastore2}
sharedDatastore={datastore1}

You still intersect 2 of these, and only pass datastore1 to CNS.

@SandeepPissay can you comment on this as well.

Reply:
What I am saying is that you maintain a map of datastores which users has permissions for instead of ignore list.

Reply:
intersect of 2 set if more expensive than my current approach.

Reply:
@SandeepPissay Do you think what Balu suggested is a must fix?

Reply:
Why? Please note that we are not intersecting sets. We still use maps only. 

This is how CreateVolume code looks like which is very simple.

    ```
var filteredDatastores []*cnsvsphere.DatastoreInfo
	for _, sharedDatastore := range sharedDatastores {
		if _, existsInDsMap := dsMap[sharedDatastore.Info.Url]; existsInDsMap {
             filteredDatastores = append(filteredDatastores, sharedDatastore)
		} else {
			log.Debugf("filter out datastore %v from create volume spec", sharedDatastore)
		}
	}
	log.Debug("filterIgnoreDatastores: filteredDatastores %v", filteredDatastores)
	return filteredDatastores
```

Reply:
Per offline discussion with @BaluDontu , I will keep the code as it is.

Reply:
Discussed offline with @lipingxue . As the testing effort is significant, I will work on these changes and raise a PR for it later on.